### PR TITLE
statistics: handle drop partition events for global-stats

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?
When drop partitions, global-stats is not automatically updated.

### What is changed and how it works?
Handle ddl events for drop partitions in global-stats.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- No release note

### CODE REVIEW

File: statistics/handle/ddl.go [hunk 1]
```diff
@@ -44,21 +46,67 @@ func (h *Handle) HandleDDLEvent(t *util.Event) error {
...
+						return err
+					}
+					for i := 0; i < newIndexGlobalStats.Num; i++ {
+						hg, cms, topN, fms := newIndexGlobalStats.Hg[i], newIndexGlobalStats.Cms[i], newIndexGlobalStats.TopN[i], newIndexGlobalStats.Fms[i]
+						err = h.SaveStatsToStorage(tableID, newIndexGlobalStats.Count, 1, hg, cms, topN, fms, 2, 1)
+						if err != nil {
+							return err
+						}
+					}
+				}
```
Comment (qw4990) [inline statistics/handle/ddl.go:108]: Could you please create a separate function for this logic?

File: statistics/handle/ddl.go [hunk 2]
```diff
@@ -44,21 +46,67 @@ func (h *Handle) HandleDDLEvent(t *util.Event) error {
...
-			// TODO: need trigger full analyze
+			// We need to merge the partition-level stats to global-stats when we drop table partition in dynamic mode.
+			tableID := t.TableInfo.ID
+			is := infoschema.GetInfoSchema(h.mu.ctx)
+			h.mu.Lock()
+			globalTable, ok := h.getTableByPhysicalID(is, tableID)
+			h.mu.Unlock()
+			if !ok {
+				err := errors.Errorf("unknown physical ID %d in stats meta table, maybe it has been dropped", tableID)
+				return err
```
Comment (qw4990) [inline statistics/handle/ddl.go:65]: What would happen if we return an error here? A warn log record?

Comment (Reminiscent) [reply statistics/handle/ddl.go:58]: Yes, a warn log record will be generated.

File: statistics/handle/handle_test.go [hunk 3]
```diff
@@ -1562,15 +1562,72 @@ partition by range (a) (
 
 	tk.MustExec("alter table t drop partition p2;")
 	c.Assert(s.do.StatsHandle().DumpStatsDeltaToKV(handle.DumpAll), IsNil)
+	tk.MustExec("analyze table t;")
 	globalStats = h.GetTableStats(tableInfo)
-	// The value of global.count will be updated the next time analyze.
-	c.Assert(globalStats.Count, Equals, int64(9))
-	c.Assert(globalStats.ModifyCount, Equals, int64(0))
```
Comment (Reminiscent) [inline statistics/handle/handle_test.go:1568]: Test this in the `TestDDLPartition4GlobalStats`.

File: statistics/handle/ddl.go [hunk 4]
```diff
@@ -44,21 +46,91 @@ func (h *Handle) HandleDDLEvent(t *util.Event) error {
...
+func (h *Handle) updateGlobalStats(tblInfo *model.TableInfo) error {
+	// We need to merge the partition-level stats to global-stats when we drop table partition in dynamic mode.
+	tableID := tblInfo.ID
+	is := infoschema.GetInfoSchema(h.mu.ctx)
+	globalStats, err := h.TableStatsFromStorage(tblInfo, tableID, true, 0)
+	if err != nil {
+		return err
+	}
+	// If we do not currently have global-stats, no new global-stats will be generated.
+	if globalStats != nil {
```
Comment (qw4990) [inline statistics/handle/ddl.go:83]: `if globalStats == nil {return nil}` is more go-style than `if globalStats != nil {...}`.

File: statistics/handle/ddl.go [hunk 5]
```diff
@@ -44,21 +46,91 @@ func (h *Handle) HandleDDLEvent(t *util.Event) error {
...
+			}
+		}
+	}
+	return nil
+}
+
+// analyzeOptionDefault saves the default values of NumBuckets and NumTopN.
+// These values will be used in dynamic mode when we drop table partition and then need to merge global-stats.
+// These values originally came from the analyzeOptionDefault structure in the planner/core/planbuilder.go file.
+var analyzeOptionDefault = map[ast.AnalyzeOptionType]uint64{
```
Comment (qw4990) [inline statistics/handle/ddl.go:68]: How about making this structure in `planbuilder.go` exposed and using it directly.

**Review by qw4990 — APPROVED** [review]

File: statistics/handle/ddl.go (hunk 5 above)
Comment (Reminiscent) [reply statistics/handle/ddl.go]: It seems will cause the `import cycle`

**Review by rebelice — APPROVED** [review]

**Pull request merged**